Rewrite the guts of torch::jit::Lexer to speed it up

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151850
* #151849
* #151810
* #151807
* #151806
* #151805
* #151804
* #151803
* #151802
* #151801

The trie-based approach was, apparently, not efficient. This incidentally fixes a bug where "not inp" and "is note" were lexed incorrectly; see test_lexer.cpp update.

Differential Revision: [D73129443](https://our.internmc.facebook.com/intern/diff/D73129443/)

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

swolchok committed:
Rewrite the guts of torch::jit::Lexer to speed it up

The trie-based approach was, apparently, not efficient.

Differential Revision: [D73129443](https://our.internmc.facebook.com/intern/diff/D73129443/)

[ghstack-poisoned]

swolchok committed:
Update on "Rewrite the guts of torch::jit::Lexer to speed it up"


The trie-based approach was, apparently, not efficient. This incidentally fixes a bug where "not inp" and "is note" were lexed incorrectly; see test_lexer.cpp update.

Differential Revision: [D73129443](https://our.internmc.facebook.com/intern/diff/D73129443/)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

Skylion007 commented on torch/csrc/jit/frontend/lexer.h:
```diff
@@ -299,8 +422,196 @@ struct TORCH_API SharedParserData {
+      int* kind,
+      StringCordView::Iterator* end) const {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(pos.has_next());
+    static char is_not[] = "is not";
```
Should be constexpr too

swolchok committed:
rebase and minor fixes as requested on "Rewrite the guts of torch::jit::Lexer to speed it up"


The trie-based approach was, apparently, not efficient. This incidentally fixes a bug where "not inp" and "is note" were lexed incorrectly; see test_lexer.cpp update.

Differential Revision: [D73129443](https://our.internmc.facebook.com/intern/diff/D73129443/)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

swolchok committed:
Update on "Rewrite the guts of torch::jit::Lexer to speed it up"


The trie-based approach was, apparently, not efficient. This incidentally fixes a bug where "not inp" and "is note" were lexed incorrectly; see test_lexer.cpp update.

Differential Revision: [D73129443](https://our.internmc.facebook.com/intern/diff/D73129443/)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

swolchok committed:
Update on "Rewrite the guts of torch::jit::Lexer to speed it up"


The trie-based approach was, apparently, not efficient. This incidentally fixes a bug where "not inp" and "is note" were lexed incorrectly; see test_lexer.cpp update.

Differential Revision: [D73129443](https://our.internmc.facebook.com/intern/diff/D73129443/)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

swolchok commented:
> I'll come back later when tests are passing

We're green!

Skylion007 commented on torch/csrc/jit/frontend/parser_constants.h:
```diff
@@ -1,5 +1,6 @@
 
 namespace torch::jit {
-static const char* valid_single_char_tokens = "+-*/%@()[]:,={}><.?!&^|~";
+[[maybe_unused]] static const char* valid_single_char_tokens =
```
```suggestion
[[maybe_unused]] static constexpr char* valid_single_char_tokens =

swolchok replied:
out of scope for this change

swolchok commented on torch/csrc/jit/frontend/lexer.h:
```diff
@@ -248,41 +209,203 @@ struct TORCH_API SharedParserData {
+            return true;
+          }
+        }
+        goto single_char_token;
```
AIUI it's typical to need `goto` for exactly the task of implementing a lexer via a state machine. for example see the code generated by re2c: https://github.com/skvadrik/re2c/blob/master/examples/c/real_world/cxx98.c

swolchok commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "This codev PR is breaking  on it's internal counterpart diff D73129443.  For codev PRs like this one, please always make sure the internal diff is green and then land the diff internally. The Github PR will be automatically merged"

cyyever commented on test/cpp/jit/test_lexer.cpp:
```diff
@@ -45,7 +45,7 @@ TEST(LexerTest, SlightlyOffIsNot) {
   std::vector<std::string> suffixes = {"", " ", "**"};
   for (const auto& suffix : suffixes) {
-    std::vector<std::string> extras = {"i", "i3"};
+    std::vector<std::string> extras = {"i", "i3", "inn"};
```
Using `auto` to for `std::initializer_list`? The same to `suffixes`

swolchok replied:
it's a test, no need to micro-optimize it

swolchok commented:
> For codev PRs like this one, please always make sure the internal diff is green and then land the diff internally

@malfet this contradicts what you told me about always being able to just land via GH1

malfet commented:
@swolchok this sounds right to me. 

I guess @ZainRizvi was numerous red signals on internal diff, not sure what caused those though (could be a bad base)

swolchok commented:
> @swolchok this sounds right to me.
> 
> I guess @ZainRizvi was numerous red signals on internal diff, not sure what caused those though (could be a bad base)

I think it was just a previous version of the diff; the latest version as landed on github hadn't had time to get CI results when it was reverted

swolchok commented:
#152250 reapplies

ZainRizvi commented:
Yeah, it seems like the  diff import tooling was lagging somehow.  It only properly synchronized the diff with this PR after I did the revert here.  Apologies for the churn
